libopenage/renderer/demo/demo_7.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2025-2025 the openage authors. See copying.md for legal info.
+
+#include "demo_7.h"
+
+#include "util/path.h"
+
+#include <eigen3/Eigen/Dense>
+#include <QKeyEvent>
+
+#include "coord/tile.h"
+#include "renderer/camera/camera.h"
+#include "renderer/gui/integration/public/gui_application_with_logger.h"
+#include "renderer/opengl/window.h"
+#include "renderer/render_factory.h"
+#include "renderer/render_pass.h"
+#include "renderer/render_target.h"
+#include "renderer/resources/assets/asset_manager.h"
+#include "renderer/resources/shader_source.h"
+#include "renderer/stages/camera/manager.h"
+#include "renderer/stages/screen/render_stage.h"
+#include "renderer/stages/skybox/render_stage.h"
+#include "renderer/stages/terrain/render_entity.h"
+#include "renderer/stages/terrain/render_stage.h"
+#include "renderer/stages/world/render_entity.h"
+#include "renderer/stages/world/render_stage.h"
+#include "renderer/uniform_buffer.h"
+#include "time/clock.h"
+
+namespace openage::renderer::tests {
+
+void renderer_demo_7(const util::Path &path) {
+	// Basic setup
+	auto qtapp = std::make_shared<gui::GuiApplicationWithLogger>();
+	window_settings settings;
+	settings.width = 800;
+	settings.height = 600;
+	settings.debug = true;
+
+	auto window = std::make_shared<opengl::GlWindow>("Shader Commands Demo", settings);
+	auto renderer = window->make_renderer();
+	auto camera = std::make_shared<renderer::camera::Camera>(renderer, window->get_size());
+	auto clock = std::make_shared<time::Clock>();
+	auto asset_manager = std::make_shared<renderer::resources::AssetManager>(
+		renderer,
+		path["assets"]["test"]);
+	auto cam_manager = std::make_shared<renderer::camera::CameraManager>(camera);
+
+	auto shaderdir = path / "assets" / "test" / "shaders";
+
+	std::vector<std::shared_ptr<RenderPass>>
+		render_passes{};
+
+	// Initialize world renderer with shader commands
+	auto world_renderer = std::make_shared<renderer::world::WorldRenderStage>(
+		window,
+		renderer,
+		camera,
+		shaderdir,
+		shaderdir, // Temporarily, Shader commands config has the same path with shaders for this demo
+		asset_manager,
+		clock);
```

Comment:
Demos should be minimal, so that we can show of specific features of the renderer without having to replicate the whole rendering pipeline. Ideally, you shouldn't initialize a whole world renderer and should just show the absolute basics, i.e.:

- Loading a shader template from file
- Adding placeholder snippets
- Getting the "complete" shader code (i.e. template with inserted placeholders)
- Rendering an object with the shader to the screen (this can be a bufferless quad with no texture)

Reply:
Oh, I get it. I was thinking about test it with the `WorldRenderStage`. Since the shader template seems to be standalone and doesn't rely on `WorldRenderStage` at all, would it make more sense to move it to the `util` namespace? So we can use it not only for world shader later.

Reply:
Oh, I get it. I was thinking about testing it with the `WorldRenderStage`.